### ISSUE TITLE
Include name of deposit function in subnet node `nft_mint_event`

### ISSUE THREAD

**Issue by rafaelcr** [hirosystems/stacks-subnets#160]
When I deposit an NFT from L1 into L2 following the [NFT use case](https://github.com/hirosystems/stacks-subnets/blob/master/NFT_USE_CASE.md), the subnet node emits an event like this:
```
{
  "committed" : true,
  "type" : "nft_mint_event",
  "nft_mint_event" : {
    "raw_value" : "0x0100000000000000000000000000000005",
    "asset_identifier" : "ST2NEB84ASENDXKYGJPQW86YXQCEFEX2ZQPG87ND.simple-nft-l2::nft-token",
    "recipient" : "ST2NEB84ASENDXKYGJPQW86YXQCEFEX2ZQPG87ND",
    "value" : {
      "UInt" : 5
    }
  },
  "event_index" : 0,
  "txid" : "0x8d042c14323cfd9d31e121cc48c2c641a8db01dce19a0f6dd531eb33689dff44"
}
```
In the API, we use this event to create a synthetic Contract Call transaction that makes this NFT be minted, which requires us to set a correct contract function call.

I originally tried setting the function_name as `nft-mint?` for this transaction, but when I try to query this tx from an API endpoint I get a crash because `nft-mint?` is not part of the `simple-nft-l2` contract's ABI.

Can the subnet node also broadcast the deposit function within this event so we can process this tx? In this case, it should be `hyperchain-deposit-nft-token`.

cc @pavitthrap

**Comment by pavitthrap** [comment]
Hi @rafaelcr - sorry for the delay, I switched over to working on the main blockchain entirely. 

> The `hc-alpha` contract uses this value to call the function on the L2 contract when performing the deposit, line 74: 

Just a note of clarification, the contract merely registers the name. The node ultimately calls the deposit function itself. 

So here's what I think the approach would be. Instead of adding the deposit function name to the nft_mint_event (as this event is used broadly for all types of mints on the L2), I think it makes sense to add a novel `deposit_nft_event`, which would include the desired information. 

For both fungible tokens and NFTs, a specified deposit function is called on the L2. The location of this call is in `src/chainstate/stacks/db/blocks.rs`, in the functions `process_deposit_ft_ops` and `process_deposit_nft_ops`. In the deposit ft function, we could make the following change (and we could make a similar change for the deposit nft function): 

<img width="1231" alt="image" src="https://user-images.githubusercontent.com/14214725/200930340-a0f55107-a144-477b-ab68-32b8cefedd20.png">

**Comment by obycode** [comment]
How about creating a new trait for NFTs and FTs on the subnet which specifies a required name and signature for this function (and potentially any other functions that might be needed)? Then the API can always know the correct name for the mint function.

**Comment by rafaelcr** [comment]
Great solution @obycode !

**Comment by obycode** [comment]
I'll take over this one since with the trait change, there is very little work on the subnet side.

**Comment by obycode** [comment]
I'll add it to the boot contract I've added for #190

**Comment by obycode** [comment]
Proposed changes:
- In `register-new-ft-contract` and `register-new-nft-contract` (L1 subnet.clar), remove the `deposit-fn-name` parameter
- In `register-new-ft-contract` and `register-new-nft-contract` (L1 subnet.clar), add a `subnet-contract` parameter
- In `deposit-ft-asset` and `deposit-nft-asset` (L1 subnet.clar), remove the `subnet-contract-id` parameter, since it is no longer needed if we ensure that the L2 contract implements this new trait
- When handling the registration of a new FT or NFT contract, validate that the L2 contract is deployed and implements the trait
  * Open question: this will happen asynchronously with respect to the L1 call to `register-new-*-contract`, so how should we handle the case when the contract does not exist or does not implement the trait

**Comment by obycode** [comment]
The trait is defined as:
```clarity
(define-trait subnet-asset
  (
    ;; Process a deposit from the burnchain.
    (deposit-from-burnchain
      (
        uint       ;; asset-id (NFT) or amount (FT)
        principal  ;; recipient
      )
      (response bool uint)
    )

    ;; Burn the asset for withdrawal from the subnet.
    (burn-for-withdrawal
      (
        uint       ;; asset-id (NFT) or amount (FT)
        principal  ;; owner
      )
      (response bool uint)
    )
  )
)
```

So for handling a deposit, the function will always be `deposit-from-burnchain`.

**Comment by obycode** [comment]
This was resolved with #206.
